### PR TITLE
translate/emitter: Implement partial indexes

### PR DESCRIPTION
**Pull request by PThorpe92**
This PR adds support for partial indexes, e.g. `CREATE INDEX` with a provided predicate
```sql
CREATE UNIQUE INDEX idx_expensive ON products(sku) where price > 100;
```

The PR does not yet implement support for using the partial indexes in the optimizer.

### CODE REVIEW

File: core/translate/index.rs [hunk 1]
```diff
@@ -159,6 +206,25 @@ pub fn translate_create_index(
     // emit MakeRecord (index key + rowid) into record_reg.
     //
     // Then insert the record into the sorter
+    let mut skip_row_label = None;
+    if let Some(where_clause) = where_clause {
+        let reg = program.alloc_register();
+        let label = program.allocate_label();
+        let pr = translate_expr(
```
Comment (jussisaurio) [inline core/translate/index.rs]: any particular reason not to use `translate_condition_expr` here?

File: core/translate/emitter.rs [hunk 2]
```diff
@@ -594,8 +633,11 @@ fn emit_delete_insns(
                 start_reg,
                 num_regs,
                 cursor_id: index_cursor_id,
-                raise_error_if_no_matching_entry: true,
+                raise_error_if_no_matching_entry: index.where_clause.is_none(),
```
Comment (jussisaurio) [inline core/translate/emitter.rs:622]: could you clarify why we are both emitting the WHERE condition AND allowing a not-found entry not to raise an error?

File: core/translate/emitter.rs [hunk 3]
```diff
@@ -1671,3 +1765,42 @@ fn init_limit(
         }
     }
 }
+
+fn rewrite_where_for_update_registers(
```
Comment (jussisaurio) [inline core/translate/emitter.rs:1794]: comment this function

File: core/translate/update.rs [hunk 4]
```diff
@@ -390,20 +391,20 @@ pub fn prepare_update_plan(
 
                 if !needs {
                     if let Some(w) = &idx.where_clause {
-                        // Bind once so names→positions are resolved exactly like execution.
                         let mut where_copy = w.as_ref().clone();
                         let mut param = ParamState::disallow();
                         let mut tr =
                             TableReferences::new(table_references.joined_tables().to_vec(), vec![]);
-                        bind_and_rewrite_expr(
+                        let _ = bind_and_rewrite_expr(
```
Comment (jussisaurio) [inline core/translate/update.rs]: why ignoring error?

File: core/translate/emitter.rs [hunk 5]
```diff
@@ -1209,8 +1210,9 @@ fn emit_update_insns(
         });
 
         // Handle unique constraint
-        if index.unique {
+        if !index.unique {
```
Comment (jussisaurio) [inline core/translate/emitter.rs]: wait, why is this condition inverted now?

File: core/translate/index.rs (hunk 1 above)
Comment (PThorpe92) [reply core/translate/index.rs]: nope, I think I did the same thing for UPSERT actually :/

File: core/translate/emitter.rs (hunk 2 above)
Comment (PThorpe92) [reply core/translate/emitter.rs:622]: ahh this was part of a "try random shit" debugging effort and I forgot to change it back

File: core/translate/emitter.rs (hunk 5 above)
Comment (PThorpe92) [reply core/translate/emitter.rs]: I typed out this whole explanation of how we now have _two_ loops, and one we used to break out of, etc. etc... just to look closely and realize that it is in fact wrong.

**Comment by jussisaurio** [comment]
could you add partials to index_mutation_fuzz or whatever the hell it is called? or make a copypaste variation of that test for partial indexes? might reveal some edge cases still

**Comment by PThorpe92** [comment]
### EDIT:
added `partial_index_mutation_and_upsert_fuzz` and it found a ton of problems. quite a bit of fixes were needed for even the existing upsert translation. (which is why this PR is now 2x the original size)

## EDIT AGAIN:
found more problems by extending the fuzzer. fixing

File: core/translate/emitter.rs [hunk 6]
```diff
@@ -561,17 +565,38 @@ fn emit_delete_insns(
...
             .unwrap_or_default();
 
         for (index, index_cursor_id) in other_indexes {
+            let skip_delete_label = if index.where_clause.is_some() {
+                let where_copy = index
+                    .bind_where_expr(Some(table_references), connection)
+                    .expect("where clause to exist");
+                let skip_label = program.allocate_label();
+                let reg = program.alloc_register();
+                translate_expr_no_constant_opt(
```
Comment (PThorpe92) [inline core/translate/emitter.rs:583]: @jussisaurio NOTE: when I made each of these into `translate_condition_expr`, they caused the fuzzer to hang indefinitely and I was so sick of fixing other bugs that the fuzzer found that I did not investigate further.

**Review by jussisaurio — APPROVED** [review]
